apps/rush-lib/src/cli/scriptActions/BulkScriptAction.ts

```diff
@@ -169,44 +185,105 @@ export class BulkScriptAction extends BaseScriptAction {
       ignoreMissingScript: this._ignoreMissingScript,
       ignoreDependencyOrder: this._ignoreDependencyOrder,
       packageDepsFilename: Utilities.getPackageDepsFilenameForCommand(this._commandToRun)
-    });
-
-    // Register all tasks with the task collection
-    const taskCollection: TaskCollection = taskSelector.registerTasks();
+    };
 
-    const taskRunner: TaskRunner = new TaskRunner(taskCollection.getOrderedTasks(), {
+    const taskRunnerOptions: ITaskRunnerOptions = {
       quietMode: isQuietMode,
       parallelism: parallelism,
       changedProjectsOnly: changedProjectsOnly,
       allowWarningsInSuccessfulBuild: this._allowWarningsInSuccessfulBuild
-    });
+    };
 
-    try {
-      await taskRunner.executeAsync();
+    const executeOptions: IExecuteInternalOptions = {
+      taskSelectorOptions,
+      taskRunnerOptions,
+      stopwatch,
+      terminal
+    };
 
-      stopwatch.stop();
-      console.log(colors.green(`rush ${this.actionName} (${stopwatch.toString()})`));
+    if (this._watchParameter.value) {
+      await this.runWatch(executeOptions);
+    } else {
+      await this._runOnce(executeOptions);
+    }
+  }
 
-      this._doAfterTask(stopwatch, true);
-    } catch (error) {
-      stopwatch.stop();
+  /**
+   * Runs the command in watch mode. Fundamentally is a simple loop:
+   * 1) Wait for a change to one or more projects in the selection (skipped initially)
+   * 2) Invoke the command on the changed projects, and, if applicable, downstream projects
+   * 3) Goto (1)
+   */
+  protected async runWatch(options: IExecuteInternalOptions): Promise<void> {
```

Comment:
If this is `protected`, you need to export `IExecuteInternalOptions` so child classes can access it.

Reply:
Switched to private for now.